sql/src/main/java/io/crate/execution/engine/window/WindowBatchIterator.java

```diff
@@ -0,0 +1,169 @@
+ * agreement.
+ */
+
+package io.crate.execution.engine.window;
+
+import io.crate.analyze.WindowDefinition;
+import io.crate.data.BatchIterator;
+import io.crate.data.CollectingRowConsumer;
+import io.crate.data.InMemoryBatchIterator;
+import io.crate.data.MappedForwardingBatchIterator;
+import io.crate.data.Row;
+import io.crate.data.RowN;
+import io.crate.data.SentinelRow;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collector;
+
+/**
+ * BatchIterator that computes an aggregate or window function against a window over the source batch iterator.
+ * Executing a function over a window means having to range over a set of rows from the source iterator. By default (ie.
+ * empty window) the range is between unbounded preceding and current row. This means the function will need to be
+ * computed against all the rows from the beginning of the partition and through the current row's _last peer_. The
+ * function result will be emitted for every row in the window.
+ * <p>
+ *      For eg.
+ * <p>
+ *      source : [ 1, 2, 3, 4 ]
+ * <p>
+ *      function : sum
+ * <p>
+ *      window definition : empty (equivalent to an empty OVER() clause)
+ * <p>
+ *      window range is all the rows in the source as all rows are peers if there is no ORDER BY in the OVER clause
+ * <p>
+ *      Execution steps :
+ * <p>
+ *          1. compute the function while the rows are peers
+ * <p>
+ *                  1 + 2 + 3 + 4 = 10
+ * <p>
+ *          2. for each row in the window emit the function result
+ * <p>
+ *                  [ 10, 10, 10, 10 ]
+ */
+public class WindowBatchIterator extends MappedForwardingBatchIterator<Row, Row> {
+
+    private final BatchIterator<Row> source;
+    private final Collector<Row, ?, Iterable<Row>> collector;
+    private WindowDefinition windowDefinition;
+
+    private Row currentElement;
+
+    private int sourceRowsConsumed;
+    private int windowRowPosition;
+    private List<Row> peerRows = new ArrayList<>();
+    private boolean peersCollected = false;
+    private Row prevRow = null;
+
+    WindowBatchIterator(WindowDefinition windowDefinition,
+                        BatchIterator<Row> source,
+                        Collector<Row, ?, Iterable<Row>> collector) {
+        assert windowDefinition.partitions().size() == 0 : "Only empty OVER() is supported now";
+        assert windowDefinition.orderBy() == null : "Only empty OVER() is supported now";
+        assert windowDefinition.windowFrameDefinition() == null : "Only empty OVER() is supported now";
+
+        this.windowDefinition = windowDefinition;
+        this.source = source;
+        this.collector = collector;
+    }
+
+    @SuppressWarnings("unused")
+    private boolean arePeers(@Nullable Row prevRow, Row nextRow) {
+        // all rows are peers when orderBy is missing
+        return windowDefinition.orderBy() == null;
```

Comment:
Maybe we should always return `false` for now? This logic isn't going to work out once orderBy is in place?

Reply:
I think the logic could mostly be reused as the thing to add would be an initial "previous" state which we build on top (so the aggregation for the next "batch" of peers is built on an aggregation that has as the initial state the collected state for the previous row)

happy to always return `true` now as well though, if that's clearer in any way